python/ray/dashboard/subprocesses/handle.py

```diff
@@ -0,0 +1,244 @@
+    SubprocessModule,
+    SubprocessModuleConfig,
+    run_module,
+)
+from ray.dashboard.subprocesses.utils import assert_not_in_asyncio_loop, ThreadSafeDict
+
+"""
+This file contains code run in the parent process. It can start a subprocess and send
+messages to it.
+"""
+
+
+class SubprocessModuleHandle:
+    """
+    A handle to a module created as a subprocess. Can send messages to the module and
+    receive responses. On destruction, the subprocess is terminated.
+    """
+
+    @dataclass
+    class ActiveRequest:
+        request: aiohttp.web.Request
+        response_fut: Awaitable[aiohttp.web.Response]
+        # Only exists when the module decides this is a streaming response.
+        # To keep the data sent in order, we use future to synchronize. This assumes
+        # the Messages received from the Queue are in order.
+        # StreamResponseStartMessage expects this to be None. It creates the future,
+        # and in async, prepares a StreamResponse and resolves the future.
+        # StreamResponseDataMessage expects a future. It *replaces* the future with a
+        # new future by a coroutine that awaits the previous future, writes the data and
+        # resolves the new future.
+        # StreamResponseEndMessage expects a future. It resolves the future and sets
+        # the stream_response to None.
+        stream_response: Optional[
+            concurrent.futures.Future[aiohttp.web.StreamResponse]
+        ] = None
+
+    def __init__(
+        self,
+        loop: asyncio.AbstractEventLoop,
+        module_cls: type[SubprocessModule],
+        config: SubprocessModuleConfig,
+    ):
+        self.loop = loop
+        self.module_cls = module_cls
+        self.config = config
+        self.child_bound_queue = multiprocessing.Queue()
+        self.parent_bound_queue = multiprocessing.Queue()
+
+        self.dispatch_parent_bound_messages_thread = None
+
+        self.active_requests = ThreadSafeDict[
+            int, SubprocessModuleHandle.ActiveRequest
+        ]()
+        self.next_request_id = 0
+
+        self.process = multiprocessing.Process(
+            target=run_module,
+            args=(
+                self.child_bound_queue,
+                self.parent_bound_queue,
+                self.module_cls,
+                self.config,
+            ),
+            daemon=True,
+        )
+        self.process.start()
+
+    def start_dispatch_parent_bound_messages_thread(self):
+        self.dispatch_parent_bound_messages_thread = threading.Thread(
+            target=dispatch_parent_bound_messages,
+            args=(self.loop, self.parent_bound_queue, self),
+            daemon=True,
+        )
+        self.dispatch_parent_bound_messages_thread.start()
+
+    def send_message(self, message: ChildBoundMessage):
+        self.child_bound_queue.put(message)
+
+    async def send_request(
+        self, method_name: str, request: Optional[aiohttp.web.Request]
+    ) -> Awaitable[aiohttp.web.Response]:
```

Comment:
Does `aiohttp.web.Server` cancel the asyncio task when a client disconnects? If so we need to ensure we handle that gracefully here